index.md

```diff
@@ -4,6 +4,145 @@
+to their final location. This has the downside of potentially doubling the
+bandwidth requirements for the file copy, depending on the S3 implementation.
+
+Alternatively, the pods could write the files directly to their final
+destination in the datastore, but this increases the chances of consistency
+problems between the database and the files on disk.
+
+#### Chained collection update
+The name of the collection can be transferred in JSON as a plain string, and
+the writer service can use the same function that the pods are currently
+using to update the database.
+
+### Kafka setup
+Phalanx with the Strimzi operator would be used to install Kafka into the
+existing Butler development and production USDF vClusters.  A Kafka topic and
+username for Prompt Processing would be added.  A retention period would be
+configured on the topic to retain messages for a certain amount of days to allow
+for events to be processed if the Butler database or the writer service is down.
+
+If we partition the Kafka topic for scalability, we need to be careful with the
+event ordering because Kafka only guarantees a consistent ordering within each
+partition, not across partitions. One option would be to partition by a key
+derived from Butler metadata (e.g. detector number).  Alternatively, if there 
+are no dependencies between events we can use round-robin partioning -- sending
+dimension records, dataset records, and the collection chain update together in
+a single event would guarantee that for the immediate use case.
+
+For communication with Prompt Processing and other applications a load balancer
+with DNS name would be added.
+
+## Fringe Benefits
+As a side-effect, Kafka orders the events in a way that makes it easy to create
+a log of the data products that were generated.  This may make it
+easier to incrementally populate downstream databases, when data is moved out
+of the embargo rack and made available to end-users.
+
+The retention period on the Prompt Processing Kakfa topic can also be used as a
+transaction log to replay event processing in the event of database corruption.
+
+## Disadvantages to this approach
+Adding an external service will complicate build and deployment.  It also makes
+it harder for unit tests to fully cover the process of writing the outputs,
+since it is split between services and adds additional infrastructure
+requirements.
+
+## The other half of the problem
+Removing the writes to the database for the outputs will not fully remove the
+need for Prompt Processing pods to communicate with the central Butler
+database.  The database is also queried for inputs needed by the pipelines
+executed on the pods.
+
+However, with writes eliminated, we no longer need to connect to the primary
+database.  We can set up a dedicated read replica for Prompt Processing,
+insulating time-critical processing from other users/services that may
+consume resources unpredictably on the primary database.
```

Comment:
Worth mentioning that this won't blind us to real-time updates to the repo, as discussed on DM-49293?